File: Java/DecodeString/DecodeString.java
import java.util.Stack;
/**
 * (4ms)
 * 这个题目最基本的想法就是使用两个栈，一个存储重复的次数，另一个存储需要重复的字符串。唯一需要注意的是重复
 * 字符串的入栈形态及其时机。
 * @author Phoenix-Z
 *
 */
public class DecodeString {

	public static String decodeString(String s) {
		//使用StringBuilder以避免生成大量的String对象
		StringBuilder ans = new StringBuilder();
		//dup用来存储重复的次数
		Stack<Integer> dup = new Stack<>();
		//strs用来存储重复的字符串
		Stack<StringBuilder> strs = new Stack<>();
		for(int i = 0; i < s.length(); i++) {
			if(Character.isDigit(s.charAt(i))) {
				int count = 0;
				while(Character.isDigit(s.charAt(i))) {
					count = count * 10 + (s.charAt(i++) - '0');
				}
				dup.push(count);
				i--;
			} else if(s.charAt(i) == '[') {
				strs.push(ans);
				ans = new StringBuilder();
			} else if(s.charAt(i) == ']') {
				StringBuilder dupString = strs.pop();
				int dupCount = dup.pop();
				while(dupCount-- > 0) {
					dupString.append(ans);
				}
				ans = dupString;
			} else {
				ans.append(s.charAt(i));
			}
		}
		return ans.toString();
    }
	
	public static void main(String[] args) {
		String s = "3[a2[c]s]2[b]";
		System.out.println(decodeString(s));
	}
}

File: Python/DecodeString/DecodeString.py
#coding:utf-8
# (35ms)
# 这个方法使用正则表达式的做法，由于Python对‘+’以及‘*’都进行的重载，所以对于3[a]2[bc]
# 而言，它就相当于3*('a')+2*('bc'),对于3[a2[c]],相当于3*('a' + 2 * ('c')),所以这个方法
# 的目标就是将原字符串使用正则表达式改成Python接受的形式。
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        while '[' in s:
            s = re.sub(r'(\d+)\[([a-z]*)\]', lambda m: int(m.group(1)) * m.group(2), s)
        return s

# (48ms)	
# 这个方法与java的思路基本一致，就是用栈来存储重复次数和重复字符串，由于Python的list
# 可以放置不同的数据类型，所以可以只使用一个栈，当然需要注意push和pop的顺序	
class Solution2(object):
    def decodeString(self, s):
		stack = []
		ans = ''
		count = 0
		for c in s:
			if c.isdigit():
				count = count * 10 + int(c)
				continue
			elif c == '[':
				stack.append(count)
				stack.append(ans)
				ans = ''
				count = 0
			elif c == ']':
				ans = stack.pop() + ans * stack.pop()
			else:
				ans += c
		return ans
		
print Solution2().decodeString("3[a]2[bc]")
print Solution2().decodeString("3[a2[c]]")
print Solution2().decodeString("2[abc]3[cd]ef")